extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/SmallRyeFaultToleranceProcessor.java

```diff
@@ -260,6 +220,67 @@ public void transform(TransformationContext ctx) {
         });
     }
 
+    @BuildStep
+    // Add reflective access to fallback methods, use bean index since fallback methods can be in bean classes
+    void processFallbackMethodsAndClases(BeanArchiveIndexBuildItem beanArchiveIndexBuildItem,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
+            BuildProducer<ReflectiveMethodBuildItem> reflectiveMethod) {
+        IndexView index = beanArchiveIndexBuildItem.getIndex();
+        Map<DotName, Set<String>> classesToScan = new HashMap<>();
+
+        for (AnnotationInstance annotation : index.getAnnotations(DotNames.FALLBACK)) {
+            AnnotationValue fallbackMethodValue = annotation.value("fallbackMethod");
+            if (fallbackMethodValue == null) {
+                continue;
+            }
+            String fallbackMethod = fallbackMethodValue.asString();
+
+            // @Fallback can only be present on methods, so this is just future-proofing
+            AnnotationTarget target = annotation.target();
+            final ClassInfo clazz = target.asMethod().declaringClass();
+            if (target.kind() != Kind.METHOD) {
+                continue;
+            }
+
+            // Scan both the hierarchy of the declaring class and its interfaces like in
+            // io.smallrye.faulttolerance.internal.SecurityActions.findDeclaredMethodNames
+            DotName name = clazz.name();
+            while (name != null && !DotNames.OBJECT.equals(name)) {
+                Set<String> methods = classesToScan.computeIfAbsent(name, k -> new HashSet<>());
+                methods.add(fallbackMethod);
+                ClassInfo classInfo = index.getClassByName(name);
+                if (classInfo == null) {
+                    break;
+                }
+                name = classInfo.superName();
+            }
+            clazz.interfaceNames().forEach(it -> classesToScan.computeIfAbsent(it, k -> new HashSet<>()).add(fallbackMethod));
```

Comment:
This means that only methods from interfaces implemented directly by the bean class are registered, but not methods from interfaces implemented by superclasses of the bean class.

Reply:
Correct, looking at the smallrye code my understanding is that it only checks those, but maybe I missed something.

See https://github.com/smallrye/smallrye-fault-tolerance/blob/c0b05c97b37f5824150af2f9e4ec09ae42b93ea9/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/SecurityActions.java#L131-L207

Now that I looked at it again the code seems to indeed check only the interfaces of the bean class but the comments mention that:

```
        // if we find a matching method on the bean class or one of its superclasses or superinterfaces,
        // then we have to check that the method is either identical to or an override of a method that:
        // - is declared on a class which is a supertype of the declaring class, or
        // - is declared on an interface which implemented by the declaring class
```

which implies that the code should scan the superinterfaces of the bean class as well but it doesn't?

Reply:
Good point, that is indeed a bug in SmallRye Fault Tolerance.

Reply:
OK, then I will adjust my PR (in anticipation of this getting fixed in SmallRye FT as well). Can you please take care of the SmallRye FT issue and post the relevant issue/PR link here for reference? I could do it but I guess you already have a fork of the repo :)

Reply:
Sure, I'm fixing it at the moment.

Reply:
Fix is in https://github.com/smallrye/smallrye-fault-tolerance/pull/1045